### DISCUSSION TITLE
`json-server` not working in Github codespaces

### DISCUSSION THREAD

**Discussion by softwareguild0** [community/community#51330] [Codespaces] [answered]
### Select Topic Area

Question

### Body

Hello,

I am a beginner learning web development and I am having issues running the `json-server` package in my GitHub Codespaces. When I try to run it, the page just keeps loading and doesn't open. I have tried following the troubleshooting steps in the documentation, but I am still having the same issue.

Could you please assist me with this issue? 

Thanks in advance

**Comment by cklanac** [comment, marked as answer]
I had a similar issue using the Codespace React template. The server would either spin and then timeout, or return a Bad Gateway error. I found a [related issue here](https://github.com/orgs/community/discussions/28563) which suggested using `--host 0.0.0.0`

### Simple Answer: 

add `--host 0.0.0.0` to the JSON Server command

`json-server db.json --host 0.0.0.0`

This will allow you to access the JSON server from your browser, CURL or a UI like Postman. But depending on how you are using it, you  may see other issues like a CORS error.

I'm using `json-server` along with Create React App. Below are my notes on what I tried, what worked and what didn't work.

### Disabling CORS (does not work for me) 

Using the `--host 0.0.0.0` makes the json-server api accessible, but if you try to fetch the data within your app you'll get a CORS error. So something like this in your app `await fetch('https://username-clever-potato-1a2s3d4f5g-3001.preview.app.github.dev/posts');` will fail due to CORS. I tried disabling CORS like: `json-server db.json --no-cors true --host 0.0.0.0` but that does not work. Maybe I'm just configuring it wrong ¯\_(ツ)_/¯

### Use Simple Proxy to avoid CORS (works but breaks HMR)

If you're using Create React App, use simple [React proxy](https://create-react-app.dev/docs/proxying-api-requests-in-development/#configuring-the-proxy-manually) by adding `"proxy": "http://localhost:3001",` to `package.json`. And remove the `--host 0.0.0.0` from the json-server call.. Note: this works but it breaks HMR (Hot Module Reloading). 

`package.json`

```json
"scripts": {
    "start": "react-scripts start",
    "api": "json-server --watch ./public/data/db.json --port 3001"
  },
"proxy": "http://localhost:3001",
```

### CRA Manual Proxy and Routes (FTW)

If you're using Create React App, the best solution I've found is to configure CRA to use a [Manual Proxy](https://create-react-app.dev/docs/proxying-api-requests-in-development/#configuring-the-proxy-manually). And then configure `json-server` with [custom routes](https://www.npmjs.com/package/json-server#user-content-add-custom-routes)

- First, add `setupProxy.js` to the `src` directory. This will proxy requests to `/api/ to the target `http://localhost:3001`. If you're not using CRA, then you can do something like this in Express. See [related issue here](https://github.com/orgs/community/discussions/28563)

```js
const { createProxyMiddleware } = require('http-proxy-middleware');

module.exports = function(app) {
  app.use(
    '/api',
    createProxyMiddleware({
      target: 'http://localhost:3001',
      changeOrigin: true,
    })
  );
};
```

- Next add a script in `package.json` to run `json-server` with a routes config

```json
"scripts": {
    "start": "react-scripts start",
    "api": "json-server --watch ./public/data/db.json --port 3001 --routes routes.json"
  },
```

- And, add a `routes.json` file to the root of the project. This will map `/api/*` requests to your json-server.

```json
{
  "/api/*": "/$1"
}
```

- Run the json-server api `npm run api`
- Run the react server `npm start`

In your app, a request like `await fetch('/api/posts');` will call your local server, which will proxy it to the json-server, which maps it to the `/posts`.
